server/resp/src/main/java/org/infinispan/server/resp/operation/SetOperation.java

```diff
@@ -0,0 +1,212 @@
+   }
+
+   private static class SetOperationOptions {
+      private final List<byte[]> arguments;
+      private byte[] key;
+      private byte[] value;
+      private long expirationMs;
+      private boolean keepTtl;
+      private boolean setAndReturnPrevious;
+      private SetOperationType operationType;
+
+      public SetOperationOptions(List<byte[]> arguments) {
+         this.arguments = arguments;
+         this.key = null;
+         this.value = null;
+         this.expirationMs = -1;
+         this.keepTtl = false;
+         this.setAndReturnPrevious = false;
+         this.operationType = SetOperationType.PUT;
+         parseAndLoadOptions();
+      }
+
+      public void withKey(byte[] key) {
+         this.key = key;
+      }
+
+      public void withValue(byte[] value) {
+         this.value = value;
+      }
+
+      public void withReturnPrevious() {
+         this.setAndReturnPrevious = true;
+      }
+
+      public void withExpiration(long expirationMs) {
+         this.expirationMs = expirationMs;
+      }
+
+      public void withOperationType(SetOperationType operationType) {
+         this.operationType = operationType;
+      }
+
+      public void withKeepTtl() {
+         this.keepTtl = true;
+      }
+
+      public boolean isKeepingTtl() {
+         return this.keepTtl;
+      }
+
+      public boolean isReturningPrevious() {
+         return this.setAndReturnPrevious;
+      }
+
+      public SetOperationType typeOfOperation() {
+         return this.operationType;
+      }
+
+      public boolean isUsingExpiration() {
+         return expirationMs > 0 || isKeepingTtl();
+      }
+
+      private void parseAndLoadOptions() {
+         Iterator<byte[]> it = arguments.iterator();
+         while (it.hasNext()) {
+            if (key == null) {
+               withKey(it.next());
+               continue;
+            }
+
+            if (value == null) {
+               withValue(it.next());
+               continue;
+            }
+
+            // Bellow here we parse the optional arguments for the SET command:
+            //
+            // * `GET`: return the previous value with this key or nil;
+            // * `NX` or `XX`: putIfAbsent or putIfPresent. Returns nil if failed, if `GET` is present, it takes precedence.
+            //
+            // And expiration related parameters:
+            //
+            // `EX` seconds: TTL in seconds;
+            // `PX` milliseconds: TTL in ms;
+            // `EXAT` timestamp: Unix time for key expiration, seconds;
+            // `PXAT` timestamp: Unix time for key expiration, milliseconds;
+            // `KEEPTTL`: keep the key current TTL.
+            //
+            // Each of the time arguments are exclusive, only one is present at a time.
+            // All these arguments can be in any order. Expiration must be followed by the proper value.
+            String arg = new String(it.next(), StandardCharsets.US_ASCII).toUpperCase();
```

Comment:
No need to do upper case.
```
127.0.0.1:6379> set foor bar ex
(error) ERR syntax error
127.0.0.1:6379> 
```

Although I personally would rather store a static byte[] that contains the EX, PX, EXAT, PXAT and KEEPTTL bytes so we don't need to allocate any Strings and we would just utilize Arrays.equals

Reply:
This syntax error is because `ex` requires the value in the next argument. For example:

```text
127.0.0.1:6379> set key value4 ex 60 xx
OK
127.0.0.1:6379> get key
"value4"
```

Reply:
I'll try to check a way to use only the bytes directly.

Reply:
Updated to use the bytes directly. Since we know this should be ASCII, we compare them in a case-insensitive way.